### PR TITLE
EVM: Implement eth_call

### LINKED ISSUE #879: EVM: Implement `eth_call` endpoint.
## Summary
Implement the endpoint according to the Ethereum spec.

### PR DESCRIPTION
**Pull request by orkunkilic**
# Description
This PR fully implements eth_call in addition to current work. This mainly adds functionality for using previous blocks and error handling.

We now also fully support TxEnv using the same configuration with Reth excluding 4844 related fields. I will support CANCUN and 4844 in the next PR fully.

## Linked Issues
- Fixes #879 #576 #505
- Related to #502 

## Testing
Describe how these changes were tested. If you've added new features, have you added unit tests?
Both pos/neg test cases are added to integration test.

## Docs
Describe where this code is documented. If it changes a documented interface, have the docs been updated?

### CODE REVIEW

File: module-system/module-implementations/sov-evm/src/query.rs [hunk 1]
```diff
@@ -208,34 +209,41 @@ impl<C: sov_modules_api::Context> Evm<C> {
...
 
-        // TODO https://github.com/Sovereign-Labs/sovereign-sdk/issues/505
         let result = executor::inspect(evm_db, &block_env, tx_env, cfg_env).unwrap();
-        let output = match result.result {
-            revm::primitives::ExecutionResult::Success { output, .. } => output,
-            _ => todo!(),
-        };
-        Ok(output.into_data().into())
+
+        match ensure_success(result.result) {
```
Comment (bkolad) [inline module-system/module-implementations/sov-evm/src/query.rs]: Will this work: ` Ok(ensure_success(result.result)?)` ?it is a bit shorter than pattern matching.

File: module-system/module-implementations/sov-evm/src/evm/error/rpc.rs [hunk 2]
```diff
@@ -0,0 +1,642 @@
...
+
+/// Errors that can occur when interacting with the `eth_` namespace
+#[derive(Debug, thiserror::Error)]
+#[allow(missing_docs)]
+pub enum EthApiError {
+    /// When a raw transaction is empty
+    #[error("Empty transaction data")]
+    EmptyRawTransactionData,
+    #[error("Failed to decode signed transaction")]
+    FailedToDecodeSignedTransaction,
```
Comment (bkolad) [inline module-system/module-implementations/sov-evm/src/evm/error/rpc.rs:32]: We have `EmptyRawTransactionData, FailedToDecodeSignedTransaction,  FailedToRecoverSigner,` in `evm/primitive_types/RawEvmTxConversionError`. Let's replace `RawEvmTxConversionError` with `EthApiError`.

File: module-system/module-implementations/sov-evm/src/evm/primitive_types.rs [hunk 3]
```diff
@@ -30,6 +30,20 @@ impl Default for BlockEnv {
...
 }
 
+// BlockEnv from SealedBlock
+impl From<&SealedBlock> for BlockEnv {
+    fn from(block: &SealedBlock) -> Self {
+        Self {
+            number: block.header.number,
+            coinbase: block.header.beneficiary,
+            timestamp: block.header.timestamp,
+            prevrandao: Default::default(),
```
Comment (bkolad) [inline module-system/module-implementations/sov-evm/src/evm/primitive_types.rs]: If we don't support it for now let's add TODO here (preferably with an issue).

Comment (orkunkilic) [reply module-system/module-implementations/sov-evm/src/evm/primitive_types.rs]: I converted it to block.header.mix_hash, apparently we set block_env.prevrandao to mix_hash in end_slot_hook and in begin_slot_hook we set prevrandao da block hash. I don't know why we are using da root hash for these values but I converted to be consistent.

Comment (bkolad) [reply module-system/module-implementations/sov-evm/src/evm/primitive_types.rs]: We use it to get "better" randomness.

**Review by bkolad — APPROVED** [review]

**Pull request merged**